lib/src/mailmap.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright 2021 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Support for `.mailmap` files.
+
+use std::fmt::Debug;
+use std::io::{self, Read};
+
+use pollster::FutureExt;
+
+use crate::backend::Signature;
+use crate::conflicts::{materialize_tree_value, MaterializedTreeValue};
+use crate::git_backend::{signature_from_git, signature_to_git};
+use crate::op_store::WorkspaceId;
+use crate::repo::Repo;
+use crate::repo_path::RepoPath;
+
+/// Models a `.mailmap` file, mapping email addresses and names to
+/// canonical ones.
+///
+/// The syntax and semantics are as described in
+/// [`gitmailmap(5)`](https://git-scm.com/docs/gitmailmap).
+#[derive(Clone, Default)]
+pub struct Mailmap(gix::mailmap::Snapshot);
+
+impl Debug for Mailmap {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("Mailmap").field(&self.0.entries()).finish()
+    }
+}
+
+impl PartialEq for Mailmap {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.entries() == other.0.entries()
+    }
+}
+
+impl Eq for Mailmap {}
+
+impl Mailmap {
+    /// Parses a `.mailmap` file, ignoring parse errors.
+    pub fn from_bytes(bytes: &[u8]) -> Self {
+        Self(gix::mailmap::Snapshot::from_bytes(bytes))
+    }
+
+    /// Reads and parses a `.mailmap` file from a reader, ignoring parse errors.
+    pub fn from_reader<R: Read>(reader: &mut R) -> io::Result<Self> {
+        let mut bytes = Vec::new();
+        reader.read_to_end(&mut bytes)?;
+        Ok(Self::from_bytes(&bytes))
+    }
+
+    /// Returns the canonical signature corresponding to `signature`. The
+    /// timestamp will not be modified.
+    pub fn resolve(&self, signature: &Signature) -> Signature {
+        self.0
+            .try_resolve(signature_to_git(signature))
+            .map(|resolved| signature_from_git(resolved.to_ref()))
+            .unwrap_or_else(|| signature.clone())
+    }
+}
+
+/// Reads and parses the `.mailmap` file from the working copy commit of the
+/// specified workspace.
+pub fn get_wc_commit_mailmap(repo: &dyn Repo, workspace_id: &WorkspaceId) -> Mailmap {
```

Comment:
Nit: I think this should just return a `Option<Mailmap>` as it isn't relevant to all repos.

Reply:
An absent `.mailmap` is the same as an empty one, and is represented under the hood by an empty `Vec` (which does no allocation), so I’m not sure this is desirable. I had it like that previously, but every caller just did `.unwrap_or_default()`. Anything that cares more specifically about the presence and structure of the `.mailmap` file in the repository will have to poke at the tree APIs directly anyway.

Reply:
while having a `Option` makes this immediately clear, having it in the doc comment should also be enough.

Reply:
Good idea, thank you! Hopefully what I’ve pushed just now covers it. (If you can think of a better name or location for this function I’m all ears, too…)

Reply:
Basically, I'd write an empty `.mailmap` is as cheap as an empty `String` or so.

Reply:
I’ve added a comment to the type; let me know what you think. Looking up the `.mailmap` probably involves allocations sometimes (e.g. to interface with the backend or to materialize the tree), but I’ve conveyed that the actual empty `Mailmap` itself is cheap.

Reply:
I think your comment is good enough, as it makes it immediately clear why we pass around a `MailMap::default()` everywhere. It also does not imply that a lookup is cheap.